src/os_posix/os_fs.c

```diff
@@ -540,9 +627,80 @@ __posix_file_write(
               " bytes at offset %" PRIuMAX,
               file_handle->name, chunk, (uintmax_t)offset);
     }
+    WT_STAT_CONN_INCRV(session, block_byte_write_syscall, len);
     return (0);
 }
 
+/*
+ * __posix_file_write_mmap --
+ *     Write the buffer into the mapped region.
+ */
+static int
+__posix_file_write_mmap(
+  WT_FILE_HANDLE *file_handle, WT_SESSION *wt_session, wt_off_t offset, size_t len, const void *buf)
+{
+    static int remap_opportunities = 0;
+    WT_DECL_RET;
+    WT_FILE_HANDLE_POSIX *pfh;
+    WT_SESSION_IMPL *session;
+
+    session = (WT_SESSION_IMPL *)wt_session;
+    pfh = (WT_FILE_HANDLE_POSIX *)file_handle;
+
+    __wt_verbose(session, WT_VERB_WRITE,
+      "write-mmap: %s, fd=%d, offset=%" PRId64
+      ","
+      "len=%" PRIu64 ", mapped buffer: %p, mapped size = %" PRIu64 ".\n",
+      file_handle->name, pfh->fd, offset, (uint64_t)len, (void *)pfh->mmap_buf,
+      (uint64_t)pfh->mmap_size);
+
+    WT_ASSERT(session, S2C(session)->mmap);
+
+    if (!pfh->mmap_file_mappable)
+        goto syscall;
+
+    /* Indicate that we might be using the mapped area */
+    (void)__wt_atomic_addv32(&pfh->mmap_usecount, 1);
+
+    /*
+     * If the I/O falls outside of the mapped buffer, or the buffer is being resized, we defer to
+     * the regular system call.
+     */
+    if (pfh->mmap_buf != NULL && pfh->mmap_size >= (size_t)offset + len && !pfh->mmap_resizing) {
+
+        memcpy((void *)(pfh->mmap_buf + offset), buf, len);
```

Comment:
No need to cast to `void *`.

Reply:
Compilation failed for me if I didn't, because the first argument to memcpy is of type void*.

Reply:
That's weird -- we `memcpy` without casting all over the place.

OK, *shrug*.